File: booking.js
// Vercel Node.js Serverless Function
// 自动服务在 /api/booking

const FEISHU = {
  appId: 'cli_aaeacecbf47a9bc0',
  appSecret: 'UmiNOo8IHbFIb1iLwEaa8gNreIem2nVD',
  baseId: 'QXazbngDbamnwGsMjEbc58TGnDh',
  tableId: 'tblRD66BfFKmKQQl'
};

let cachedToken = null;
let tokenExpire = 0;

async function getToken() {
  if (cachedToken && Date.now() < tokenExpire - 60000) {
    return cachedToken;
  }
  const resp = await fetch('https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json; charset=utf-8' },
    body: JSON.stringify({ app_id: FEISHU.appId, app_secret: FEISHU.appSecret })
  });
  const data = await resp.json();
  cachedToken = data.tenant_access_token;
  tokenExpire = Date.now() + (data.expire || 7200) * 1000;
  return cachedToken;
}

const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'Content-Type',
  'Access-Control-Allow-Methods': 'POST, OPTIONS, GET',
  'Content-Type': 'application/json; charset=utf-8'
};

module.exports = async (req, res) => {
  // CORS
  if (req.method === 'OPTIONS') {
    res.statusCode = 204;
    Object.entries(corsHeaders).forEach(([k, v]) => res.setHeader(k, v));
    return res.end();
  }

  if (req.method === 'GET') {
    Object.entries(corsHeaders).forEach(([k, v]) => res.setHeader(k, v));
    res.statusCode = 200;
    return res.end(JSON.stringify({ status: 'ok' }));
  }

  if (req.method !== 'POST') {
    res.statusCode = 405;
    return res.end(JSON.stringify({ error: 'Method not allowed' }));
  }

  // 读取 POST body
  let body = '';
  for await (const chunk of req) body += chunk;
  let data;
  try {
    data = JSON.parse(body || '{}');
  } catch (e) {
    Object.entries(corsHeaders).forEach(([k, v]) => res.setHeader(k, v));
    res.statusCode = 400;
    return res.end(JSON.stringify({ code: -1, msg: '数据格式错误' }));
  }

  const phone = (data.phone || '').trim();
  const room = (data.roomName || '').trim();
  if (!phone || !room) {
    Object.entries(corsHeaders).forEach(([k, v]) => res.setHeader(k, v));
    res.statusCode = 400;
    return res.end(JSON.stringify({ code: -1, msg: '电话和项目不能为空' }));
  }

  try {
    const token = await getToken();
    const url = `https://open.feishu.cn/open-apis/bitable/v1/apps/${FEISHU.baseId}/tables/${FEISHU.tableId}/records`;
    const writeResp = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json; charset=utf-8'
      },
      body: JSON.stringify({
        fields: {
          '称呼': (data.name || '').trim(),
          '电话': phone,
          '项目': room,
          '日期': (data.date || '').trim(),
          '时间': (data.time || '').trim(),
          '人数': parseInt(data.people, 10) || 0,
          '备注': (data.remark || '').trim()
        }
      })
    });
    const writeData = await writeResp.json();
    if (writeData.code !== 0) {
      Object.entries(corsHeaders).forEach(([k, v]) => res.setHeader(k, v));
      res.statusCode = 500;
      return res.end(JSON.stringify({ code: -1, msg: '飞书写入失败: ' + writeData.msg }));
    }

    Object.entries(corsHeaders).forEach(([k, v]) => res.setHeader(k, v));
    res.statusCode = 200;
    return res.end(JSON.stringify({ code: 0, msg: '预约成功' }));
  } catch (e) {
    Object.entries(corsHeaders).forEach(([k, v]) => res.setHeader(k, v));
    res.statusCode = 500;
    return res.end(JSON.stringify({ code: -1, msg: e.message || String(e) }));
  }
};

File: api/booking.py
import json, time, urllib.request, ssl

FEISHU_APP_ID = 'cli_aaeacecbf47a9bc0'
FEISHU_APP_SECRET = 'UmiNOo8IHbFIb1iLwEaa8gNreIem2nVD'
BASE_ID = 'QXazbngDbamnwGsMjEbc58TGnDh'
TABLE_ID = 'tblRD66BfFKmKQQl'

_token = None
_token_expire = 0


def get_token():
    global _token, _token_expire
    now = time.time()
    if _token and now < _token_expire - 60:
        return _token
    data = json.dumps({'app_id': FEISHU_APP_ID, 'app_secret': FEISHU_APP_SECRET}).encode()
    req = urllib.request.Request(
        'https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal',
        data=data,
        headers={'Content-Type': 'application/json; charset=utf-8'}
    )
    resp = urllib.request.urlopen(req, timeout=10)
    body = json.loads(resp.read())
    _token = body['tenant_access_token']
    _token_expire = now + body.get('expire', 7200)
    return _token


def write_feishu(data):
    token = get_token()
    body = {
        'fields': {
            '称呼': (data.get('name') or '').strip(),
            '电话': (data.get('phone') or '').strip(),
            '项目': (data.get('roomName') or '').strip(),
            '日期': (data.get('date') or '').strip(),
            '时间': (data.get('time') or '').strip(),
            '人数': _to_int(data.get('people')),
            '备注': (data.get('remark') or '').strip()
        }
    }
    url = f'https://open.feishu.cn/open-apis/bitable/v1/apps/{BASE_ID}/tables/{TABLE_ID}/records'
    req = urllib.request.Request(
        url,
        data=json.dumps(body).encode('utf-8'),
        headers={
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json; charset=utf-8'
        }
    )
    resp = urllib.request.urlopen(req, timeout=10)
    return json.loads(resp.read())


def _to_int(v):
    try:
        return int(v)
    except (ValueError, TypeError):
        return 0


class Handler:
    """Vercel Python Serverless 入口"""

    def __init__(self):
        pass

    def __call__(self, request, response):
        return self.handler(request, response)

    def handler(self, request, response):
        # CORS
        response.headers.update({
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Headers': 'Content-Type',
            'Access-Control-Allow-Methods': 'POST, OPTIONS'
        })

        if request.method == 'OPTIONS':
            response.status = 204
            response.body = ''
            return response

        try:
            body = json.loads(request.body.decode('utf-8'))
        except Exception:
            response.status = 400
            response.body = json.dumps({'code': -1, 'msg': '数据格式错误'})
            return response

        phone = (body.get('phone') or '').strip()
        room = (body.get('roomName') or '').strip()
        if not phone or not room:
            response.status = 400
            response.body = json.dumps({'code': -1, 'msg': '电话和项目不能为空'})
            return response

        try:
            write_feishu(body)
            response.status = 200
            response.body = json.dumps({'code': 0, 'msg': '预约成功'}, ensure_ascii=False)
        except Exception as e:
            response.status = 500
            response.body = json.dumps({'code': -1, 'msg': str(e)}, ensure_ascii=False)

        return response


# Vercel 自动查找并注册 handler
handler = Handler()
